### PR TITLE
DI in plugins

### PR DESCRIPTION
**Pull request by BaronGreenback**
Issue
======
MVC DI, once created, cannot be changed.
The current plugins are initialised after web-server finalisation, so are excluded from the DI.

Three possible solutions exist. 1) implement a different DI, 2) write our own, or 3) implement non-dynamic plugin DI.

**Non-Dynamic Plugin DI**
This method works by splitting the plugin creation into two parts.

**Part 1:** The DI registration.  This is done through an object implementing the IPluginRegistrar interface.

Important things to note:
- Most classes are not yet instantiated at this point. However, NetworkManager, ApplicationHost, and ConfigurationManager are instantiated.
- The ConfigurationManager has not yet loaded any named configurations. Any named configurations must be manually loaded through the IConfigurationManager.RegisterConfiguration<> method, and will be re-created after DI initiation.

Implementing an object via IPluginRegistrar is optional.

**Part 2:** The Plugin instantiation - changes.

- A ConfigurationChanged has been added to better mimic how ConfigurationManager works.
- The Register and UnRegister functions have been removed from the class.
- The registration of the plugin properties have been moved from ApplicationHost.cs/LoadPlugIn, so that they are available in the constructor. (see #4241)
- CanUninstall is further restricted to non DI registered objects.

### CODE REVIEW

**Comment by crobibero** [comment]
It looks like there are some unrelated changes in this PR

**Comment by BaronGreenback** [comment]
Sorry - been playing will clean out the crap

**Comment by crobibero** [comment]
If this is the way it needs to be, I think it's the incorrect place to do it. To me the plugin DI registration should happen after `ApplicationHost.Init -> RegisterServices`.

Currently the `RegisterServices` function will be activated for every assembly.
Additionally, I believe the call to `plugin.RegisterServices(ServiceCollection);` should be removed.

**Comment by BaronGreenback** [comment]
I light of #4254  I think the entire LoadPlugIn method can be removed from this module.

And this piece of code in Init() needs to be moved to after the plugsin are actually loaded - as at the moment PlugsIn is always unassigned at this point.

```
if (Plugins != null)
            {
                var pluginBuilder = new StringBuilder();

                foreach (var plugin in Plugins)
                {
                    pluginBuilder.Append(plugin.Name)
                        .Append(' ')
                        .Append(plugin.Version)
                        .AppendLine();
                }

                Logger.LogInformation("Plugins: {Plugins}", pluginBuilder.ToString());
            }
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 1]
```diff
@@ -499,24 +501,11 @@ public void Init()
                 HttpsPort = ServerConfiguration.DefaultHttpsPort;
             }
 
-            if (Plugins != null)
```
Comment (BaronGreenback) [inline Emby.Server.Implementations/ApplicationHost.cs:502]: Moved from init as Plugins would always be null at that part.

**Comment by crobibero** [comment]
As discussed in #jellyfin-dev:matrix.org-

An interface will be created that plugins can implement.
https://matrix.to/#/!YOoxJKhsHoXZiIHyBG:matrix.org/$vouelLu7YAegdh3bS5zL7LLvMLv6IpjWbHa2hiRxCgo?via=bonifacelabs.ca&via=matrix.org&via=t2bot.io

File: MediaBrowser.Common/Plugins/IPluginRegistrar.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+namespace MediaBrowser.Common.Plugins
+{
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Defines the <see cref="IPluginRegistrar" />.
+    /// </summary>
+    public interface IPluginRegistrar
```
Comment (cvium) [inline MediaBrowser.Common/Plugins/IPluginRegistrar.cs:8]: I think this is a better name.
```suggestion
    public interface IPluginServiceRegistrator
```

File: Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs [hunk 3]
```diff
@@ -133,6 +133,35 @@ protected set
             }
         }
 
+        /// <summary>
+        /// Manually pre-loads a factory so that it is available pre system initialisation.
+        /// </summary>
+        /// <typeparam name="T">Class to register.</typeparam>
+        public virtual void RegisterConfiguration<T>()
```
Comment (cvium) [inline Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs:140]:
```suggestion
        public virtual void RegisterConfiguration<T>() where T : IConfigurationFactory
```

File: Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs [hunk 4]
```diff
@@ -133,6 +133,35 @@ protected set
...
+        {
+            if (!typeof(IConfigurationFactory).IsAssignableFrom(typeof(T)))
+            {
+                throw new ArgumentException("Parameter does not implement IConfigurationFactory");
+            }
+
```
Comment (cvium) [inline Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs:146]:
```suggestion
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 5]
```diff
@@ -499,24 +498,11 @@ public void Init()
...
-                }
-
-                Logger.LogInformation("Plugins: {Plugins}", pluginBuilder.ToString());
-            }
-
             DiscoverTypes();
 
             RegisterServices();
+
+            RegisterPlugInServices();
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs:505]:
```suggestion
            RegisterPluginServices();
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 6]
```diff
@@ -872,6 +825,22 @@ protected void DiscoverTypes()
             _allConcreteTypes = GetTypes(GetComposablePartAssemblies()).ToArray();
         }
 
+        private void RegisterPlugInServices()
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs:828]:
```suggestion
        private void RegisterPluginServices()
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 7]
```diff
@@ -872,6 +825,22 @@ protected void DiscoverTypes()
...
+            foreach (var pluginServiceRegistrar in GetExportTypes<IPluginRegistrar>())
+            {
+                try
+                {
+                    var instance = (IPluginRegistrar)Activator.CreateInstance(pluginServiceRegistrar);
+                    instance.RegisterServices(ServiceCollection);
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogError(ex, "Error registering {Assembly} with D.I.", pluginServiceRegistrar.Assembly);
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs:839]:
```suggestion
                    Logger.LogError(ex, "Error registering plugin services from {Assembly}.", pluginServiceRegistrar.Assembly);
```

File: MediaBrowser.Common/Plugins/BasePlugin.cs [hunk 8]
```diff
@@ -140,6 +132,30 @@ protected BasePlugin(IApplicationPaths applicationPaths, IXmlSerializer xmlSeria
         {
             ApplicationPaths = applicationPaths;
             XmlSerializer = xmlSerializer;
+            if (this is IPluginAssembly assemblyPlugin)
```
Comment (cvium) [inline MediaBrowser.Common/Plugins/BasePlugin.cs:135]: Why was this moved?

**Review by cvium — CHANGES_REQUESTED** [review]

File: MediaBrowser.Common/Plugins/BasePlugin.cs (hunk 8 above)
Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/BasePlugin.cs:135]: The settings of the plugin aren't available in the constructor. See #4241

File: MediaBrowser.Common/Plugins/IPluginRegistrar.cs [hunk 9]
```diff
@@ -5,7 +5,7 @@ namespace MediaBrowser.Common.Plugins
     /// <summary>
     /// Defines the <see cref="IPluginRegistrar" />.
     /// </summary>
-    public interface IPluginRegistrar
+    public interface IPluginServiceRegistrator
```
Comment (cvium) [inline MediaBrowser.Common/Plugins/IPluginRegistrar.cs:8]: btw, you manually have to rename the file and fix the references

Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/IPluginRegistrar.cs]: thanks twas late here. Just renamed.

**Review by crobibero — APPROVED** [review]

**Review by crobibero — APPROVED** [review]

**Comment by BaronGreenback** [comment]
@cvium Changes have all been made.

File: Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs [hunk 10]
```diff
@@ -133,6 +133,31 @@ protected set
...
+            if (_configurationFactories == null)
+            {
+                _configurationFactories = new[] { factory };
+            }
+            else
+            {
+                var list = _configurationFactories.ToList();
+                list.Add(factory);
+                _configurationFactories = list.ToArray();
```
Comment (cvium) [inline Emby.Server.Implementations/AppBase/BaseConfigurationManager.cs:153]:
```suggestion
                var oldLen = _configurationFactories.Length;
                var arr = new IConfigurationFactory[oldLen + 1];
                _configurationFactories.CopyTo(arr, 0);
                arr[oldLen] = factory;
                _configurationFactories = arr;
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 11]
```diff
@@ -839,6 +824,22 @@ protected void DiscoverTypes()
             _allConcreteTypes = GetTypes(GetComposablePartAssemblies()).ToArray();
         }
 
+        private void RegisterPluginServices()
+        {
+            foreach (var pluginServiceRegistrar in GetExportTypes<IPluginServiceRegistrator>())
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs:829]:
```suggestion
            foreach (var pluginServiceRegistrator in GetExportTypes<IPluginServiceRegistrator>())
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 12]
```diff
@@ -839,6 +824,22 @@ protected void DiscoverTypes()
...
+            {
+                try
+                {
+                    var instance = (IPluginServiceRegistrator)Activator.CreateInstance(pluginServiceRegistrar);
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs:833]:
```suggestion
                    var instance = (IPluginServiceRegistrator)Activator.CreateInstance(pluginServiceRegistrator);
```

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 13]
```diff
@@ -839,6 +824,22 @@ protected void DiscoverTypes()
...
+                    instance.RegisterServices(ServiceCollection);
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogError(ex, "Error registering plugin services from {Assembly}.", pluginServiceRegistrar.Assembly);
```
Comment (cvium) [inline Emby.Server.Implementations/ApplicationHost.cs:838]:
```suggestion
                    Logger.LogError(ex, "Error registering plugin services from {Assembly}.", pluginServiceRegistrator.Assembly);
```

File: MediaBrowser.Common/Plugins/BasePlugin.cs [hunk 14]
```diff
@@ -56,7 +56,8 @@ public abstract class BasePlugin : IPlugin, IPluginAssembly
         /// Gets a value indicating whether the plugin can be uninstalled.
         /// </summary>
         public bool CanUninstall => !Path.GetDirectoryName(AssemblyFilePath)
-            .Equals(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), StringComparison.InvariantCulture);
+            .Equals(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), StringComparison.InvariantCulture) &&
+            !typeof(IPluginServiceRegistrator).IsAssignableFrom(this.GetType());
```
Comment (cvium) [inline MediaBrowser.Common/Plugins/BasePlugin.cs:60]:
```suggestion
            .Equals(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), StringComparison.InvariantCulture)
            && !typeof(IPluginServiceRegistrator).IsAssignableFrom(this.GetType());
```

File: MediaBrowser.Common/Plugins/IPluginServiceRegistrator.cs [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+    /// <summary>
+    /// Defines the <see cref="IPluginServiceRegistrator" />.
+    /// </summary>
+    public interface IPluginServiceRegistrator
+    {
+        /// <summary>
+        /// Registers the plugin's services with the service collection.
+        /// This object is created prior to the plugin creation, so access to other classes is limited.
+        /// </summary>
```
Comment (cvium) [inline MediaBrowser.Common/Plugins/IPluginServiceRegistrator.cs:13]:
```suggestion
        /// </summary>
        /// <remarks>
        /// This interface is only used for service registration and requires a parameterless constructor.
        /// </remarks>
```

**Review by cvium — CHANGES_REQUESTED** [review]

**Review by cvium — APPROVED** [review]

**Comment by cvium** [comment]
Awaiting third approval

**Review by EraYaN — APPROVED** [review]

File: MediaBrowser.Common/Plugins/BasePlugin.cs [hunk 16]
```diff
@@ -56,7 +56,8 @@ public abstract class BasePlugin : IPlugin, IPluginAssembly
         /// Gets a value indicating whether the plugin can be uninstalled.
         /// </summary>
         public bool CanUninstall => !Path.GetDirectoryName(AssemblyFilePath)
-            .Equals(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), StringComparison.InvariantCulture);
+            .Equals(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), StringComparison.InvariantCulture)
+            && !typeof(IPluginServiceRegistrator).IsAssignableFrom(this.GetType());
```
Comment (Bond-009) [inline MediaBrowser.Common/Plugins/BasePlugin.cs:60]: Why?

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.Common/Plugins/BasePlugin.cs (hunk 16 above)
Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/BasePlugin.cs:60]: Because as it stands, plugins using this method of DI cannot be uninstalled, else they would leave a "hanging" pointer in the DI subsystem, as i haven't found a way to unregister them.

Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/BasePlugin.cs:60]: Unless, of course, uninstalled means "uninstalled on next restart", in which case the additional line can be removed.

Comment (Bond-009) [reply MediaBrowser.Common/Plugins/BasePlugin.cs:60]: Uninstalling plugins already always requires a restart

Comment (BaronGreenback) [reply MediaBrowser.Common/Plugins/BasePlugin.cs]: removed.

**Review by Bond-009 — APPROVED** [review]
